src/device/cpu/riscv_rv32ima/tlb.c

```diff
@@ -0,0 +1,224 @@
+        index_ktlb(tlb, vpn).asid = asid;
+        index_ktlb(tlb, vpn).valid = true;
+    }
+}
+
+/** Retrieves a cached mapping
+ * gives priority to megapage mappings
+ */
+extern bool rv_tlb_get_mapping(rv_tlb_t* tlb, unsigned asid, uint32_t virt, sv32_pte_t* pte, bool* megapage){
+    uint32_t vpn = virt >> 12;
+    uint32_t mvpn = vpn >> 10;
+
+    // Megapages have priority
+    if(index_mtlb(tlb, mvpn).valid && index_mtlb(tlb, mvpn).vpn == mvpn){
+        if(index_mtlb(tlb, mvpn).global || index_mtlb(tlb, mvpn).asid == asid){
+            *pte = index_mtlb(tlb, mvpn).pte;
+            *megapage = true;
+            return true;
+        }
+    }
+
+    if(index_ktlb(tlb, vpn).valid && index_ktlb(tlb, vpn).vpn == vpn){
+        if(index_ktlb(tlb, vpn).global || index_ktlb(tlb, vpn).asid == asid){
+            *pte = index_ktlb(tlb, vpn).pte;
+            *megapage = false;
+            return true;
+        }
+    }
+    
+    return false;
+}
+
+/** TLB flushes */
+
+// Invalidates all entries
+extern void rv_tlb_flush(rv_tlb_t* tlb){
+    for(size_t i = 0; i < tlb->ktlb_size; ++i){
+        tlb->ktlb[i].valid = false;
+    }
+    for(size_t i = 0; i < tlb->mtlb_size; ++i){
+        tlb->mtlb[i].valid = false;
+    }
+}
+
+// Invalidates all entries of the given asid
+extern void rv_tlb_flush_by_asid(rv_tlb_t* tlb, unsigned asid){
+    for(size_t i = 0; i < tlb->ktlb_size; ++i){
+        if(tlb->ktlb[i].global)
+            continue;
+        if(tlb->ktlb[i].asid == asid)
+            tlb->ktlb[i].valid = false;
+    }
+    for(size_t i = 0; i < tlb->mtlb_size; ++i){
+        if(tlb->mtlb[i].global)
+            continue;
+        if(tlb->mtlb[i].asid == asid)
+            tlb->mtlb[i].valid = false;
+    }
+}
+
+// Invalidates all entries that map the given virtual address
+extern void rv_tlb_flush_by_addr(rv_tlb_t* tlb, uint32_t virt){
+    uint32_t vpn = virt >> RV_PAGESIZE;
+    uint32_t mvpn = virt >> RV_MEGAPAGESIZE;
+    
+    index_ktlb(tlb, vpn).valid = false;
+    index_mtlb(tlb, mvpn).valid = false;
+}
+
+// Invalidates all entries that map the given address and are of the given asid
+extern void rv_tlb_flush_by_asid_and_addr(rv_tlb_t* tlb, unsigned asid, uint32_t virt){
+    uint32_t vpn = virt >> RV_PAGESIZE;
+    uint32_t mvpn = virt >> RV_MEGAPAGESIZE;
+    
+    // Flush from KTLB if the mapping is not global and the asid matches
+    if(!index_ktlb(tlb, vpn).global && index_ktlb(tlb, vpn).asid == asid)
+        index_ktlb(tlb, vpn).valid = false;
+    
+    // Flush from MTLB if the mapping is not global and the asid matches
+    if(!index_mtlb(tlb, mvpn).global && index_mtlb(tlb, mvpn).asid == asid)
+        index_mtlb(tlb, mvpn).valid = false;
+}
+
+/** Initializes the TLB data structure */
+extern bool rv_tlb_init(rv_tlb_t* tlb, size_t ktlb_size, size_t mtlb_size){
+    if(ktlb_size == 0 || !IS_POWER_OF_2(ktlb_size) || mtlb_size == 0 || !IS_POWER_OF_2(mtlb_size)){
+        return false;
```

Comment:
Do we expect that this would be actually called with wrong parameters? Shouldn't we assert instead?

Reply:
No we probably do not.
Changed it to an assert and removed the power of 2 requirement as it is not needed by the implementation.
Also changed the return type to void.
(all in [d8210af](https://github.com/d-iii-s/msim/pull/36/commits/d8210afea32d8c76b88054151fed5de807a85116) )